### ISSUE TITLE
filter_by with secondary, lazy and polymorphic relation raise error

### ISSUE THREAD

**Issue by Sheile** [sqlalchemy/sqlalchemy#7868]
### Describe the bug

`filter` with following relationship does work correctly, but `filter_by` does not work.
Almost the same logic was working correctly with SQLAlchemy 1.3.24.

IMO, #7239 is related.

### To Reproduce

```python
from sqlalchemy import create_engine
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class GrandParent(Base):
    __tablename__ = 'grand_parents'
    id = Column(Integer, primary_key=True)

    children = relationship('Parent')
    grand_children = relationship('Child', secondary='parents', lazy='dynamic', viewonly=True)


class Parent(Base):
    __tablename__ = 'parents'
    id = Column(Integer, primary_key=True)
    grand_parent_id = Column(Integer, ForeignKey('grand_parents.id'), nullable=False)

    children = relationship('Child')


class Child(Base):
    __tablename__ = 'children'
    id = Column(Integer, primary_key=True)
    sex = Column(String)
    parent_id = Column(Integer, ForeignKey('parents.id'), nullable=False)

    __mapper_args__ = {
        'polymorphic_on': sex,
        'polymorphic_identity': 'unknown',
        'with_polymorphic': '*'
    }


class Boy(Child):
    __tablename__ = 'boys'
    id = Column(Integer, ForeignKey('children.id'), primary_key=True)

    __mapper_args__ = {
        'polymorphic_identity': 'male',
    }


class Girl(Child):
    __tablename__ = 'girls'
    id = Column(Integer, ForeignKey('children.id'), primary_key=True)

    __mapper_args__ = {
        'polymorphic_identity': 'female',
    }


engine = create_engine('sqlite://', echo=True, future=True)
Base.metadata.create_all(engine)

with Session(engine) as session:
    grand_parent = GrandParent()
    parent = Parent()
    child = Boy()

    grand_parent.children.append(parent)
    parent.children.append(child)

    session.add(grand_parent)
    session.commit()


with Session(engine) as session:
    grand_parent = session.query(GrandParent).get(1)

    # Success
    child = grand_parent.grand_children.filter(Child.id == 1).first()

    # Failed
    child = grand_parent.grand_children.filter_by(id=1).first()
```

### Error

```
Traceback (most recent call last):
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/sql/base.py", line 1222, in __getattr__
    return self._index[key]
KeyError: 'id'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/sql/base.py", line 1693, in _entity_namespace_key
    return getattr(ns, key)
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/sql/base.py", line 1224, in __getattr__
    util.raise_(AttributeError(key), replace_context=err)
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
AttributeError: id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "reproduce.py", line 82, in <module>
    child = grand_parent.grand_children.filter_by(id=1).first()
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 1795, in filter_by
    for key, value in kwargs.items()
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 1795, in <listcomp>
    for key, value in kwargs.items()
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/sql/base.py", line 1700, in _entity_namespace_key
    replace_context=err,
  File "/home/centos/.pyenv/versions/sqlalchemy-issue/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
sqlalchemy.exc.InvalidRequestError: Entity namespace for "children LEFT OUTER JOIN boys ON children.id = boys.id LEFT OUTER JOIN girls ON children.id = girls.id" has no property "id"
``` 

### Versions

- OS: CentOS7
- Python: 3.7.4
- SQLAlchemy: 1.4.32
- Database: PostgreSQL12, SQLite
- DBAPI: psycopg

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**use annotated entity when adding secondary** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3752

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**use annotated entity when adding secondary** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3753

**Comment by zzzeek** [comment]
thanks for the clear test.   all the elements you have above are necessary to illustrate the bug :)

**Comment by zzzeek** [comment]
wasn't related to #7239 though, it's the way that #4349 got ported to 1.4

**Comment by Sheile** [comment]
Thank you for quick reply and fix!

Our products is working correctly with locally patched the SQLAlchemy 1.4.
I'm looking forward to be released this.
